src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,254 @@
+   thread access to its `Waker` and the other thread could wake it up. This
+   could happen while the `poll` is executing, before `poll` returns
+   `Poll::Pending`. Therefore, Barbara concludes that if `wake` is called while
+   a future is being polled then the future should be re-polled, even if the
+   current `poll` returns `Poll::Pending`. This requires putting a retry loop
+   into the executor.
+1. A kernel callback may call `Waker::wake` after its future returns
+   `Poll::Ready`. After `poll` returns `Poll::Ready`, the executor should not
+   `poll` the future again, so Barbara adds code to ignore those wakeups. This
+   duplicates the "ignore spurious wakeups" functionality that exists in the
+   future itself.
+
+Ultimately, this made the executor logic [quite
+nontrivial](https://github.com/tock/design-explorations/blob/master/size_comparison/futures/src/task.rs).
+The executor logic can be monomorphized for each future, which allows the
+compiler to make inlining optimizations, but results in a significant amount of
+duplicate code. Alternatively, it could be adapted to use function pointers or
+vtables to avoid the code duplication, but then the compiler *definitely* cannot
+inline `Future::poll` into the kernel callbacks.
+
+At this point, Barbara realizes she needs to do some benchmarking. Barbara comes
+up with a sample application -- an app that blinks a led and responds to button
+presses -- and implements it twice. One implementation does not use `Future` at
+all, the other does. In addition to two async drivers (a timer driver and a GPIO
+driver), the `Future`-based app uses Barbara's executor as well as a single
+combinator (that manages the state of the app).
+
+Barbara publishes an
+[analysis](https://github.com/tock/design-explorations/tree/master/size_comparison)
+of the relative sizes of the two app implementations, finding a large percentage
+increase in both code size and RAM usage (note: stack usage was not
+investigated). Most of the code size increase is from the future combinator
+code. There are two kernel callbacks in use (one that notifies the app of a
+button press and one that notifies it of a timer expiration).
+
+In the no-`Future` version of the app, a kernel callback causes the following:
+
+1. The kernel callback calls the application logic's event-handling function for
+   the specific event type.
+2. The application handles the event.
+
+The call in step 1 is inlined, so the compiled kernel callback consists only of
+the application's event-handling logic.
+
+In the `Future`-based version of the app, a kernel callback causes the
+following:
+
+1. The kernel callback updates some global state to indicate the event happened.
+2. The kernel callback invokes `Waker::wake`.
+3. `Waker::wake` calls `poll` on the application future.
+4. The application future has to look at the state saved in step 1 to determine
+   what event happened.
+5. The application future handles the event.
+
```

Comment:
But the `Future`s based version provides a lot more guarantees! E.g. it avoids reentrancy problems and makes sure all state associated with the async operation is only modified in a single place (and from a single thread/task). It's a bit of an apples to oranges comparison. If a trivial callback is good enough then `Future`s will likely have overhead. But the places where state has to be persisted in a complex fashion across  `.await` pointers is where `Future`s win, since they don't require encoding of complex state machines.

Reply:
Avoiding reentrancy problems, making async state management easy, and easily representing complex asynchronous operations are all *really* nice features of `Future`! I don't contest that.

However, they are all nice-to-have features, and it takes an awful lot of nice-to-have features to outweigh one deal-breaker. The amount of code generated by the executor and in `poll` that doesn't exist in the non-`Future` version is probably a deal-breaker for us.